### PR TITLE
Fixed `WandbLogger` `save_dir` is not set after creation (#12748)

### LINKED ISSUE #12748: WandbLogger `save_dir` should be set after creation
## 🐛 Bug

Save dir is not defined on instantiation as the docs suggest: https://github.com/PyTorchLightning/pytorch-lightning/blob/16b958095849e5c03805918d66bf4a7e1d32da44/pytorch_lightning/loggers/wandb.py#L221 https://github.com/PyTorchLightning/pytorch-lightning/blob/16b958095849e5c03805918d66bf4a7e1d32da44/pytorch_lightning/loggers/wandb.py#L305

### To Reproduce

```python
from pytorch_lightning.loggers import WandbLogger

logger = WandbLogger(project="foo")
assert logger.save_dir is not None
```

### Expected behavior

We want this because the `LightningCLI` tries to access the `log_dir` https://github.com/PyTorchLightning/pytorch-lightning/blob/16b958095849e5c03805918d66bf4a7e1d32da44/pytorch_lightning/utilities/cli.py#L430-L432 with https://github.com/PyTorchLightning/pytorch-lightning/blob/16b958095849e5c03805918d66bf4a7e1d32da44/pytorch_lightning/trainer/trainer.py#L2186 but it's not available on creation

### Additional context

This can also be reproduced with

```python
from pytorch_lightning.utilities.cli import LightningCLI
from tests.helpers import BoringModel

cli = LightningCLI(BoringModel, run=False, trainer_defaults={"max_epochs": 1})
cli.trainer.fit(cli.model)
print(cli.trainer.logger)
```

```bash
$ python bug.py --trainer.logger=WandbLogger
...
Traceback (most recent call last):
  File "/home/carmocca/git/pytorch-lightning/kk.py", line 5, in <module>
    cli.trainer.fit(cli.model)
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 724, in fit
    self._call_and_handle_interrupt(
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 677, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 765, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 1128, in _run
    self._call_setup_hook()  # allow user to setup lightning_module in accelerator environment
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 1447, in _call_setup_hook
    self._call_callback_hooks("setup", stage=fn)
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 1590, in _call_callback_hooks
    fn(self, self.lightning_module, *args, **kwargs)
  File "/home/carmocca/git/pytorch-lightning/pytorch_lightning/utilities/cli.py", line 432, in setup
    assert log_dir is not None
```

cc @awaelchli @morganmcg1 @AyushExel @borisdayma @scottire @manangoel99

### PR DESCRIPTION
**Pull request by tshu-w**
## What does this PR do?

Fixes #12748 make sure `WandbLogger` `save_dir` set after creation.

### Does your PR introduce any breaking changes? If yes, please list them.

No.

### Test

```python
from pytorch_lightning.cli import LightningCLI

from pytorch_lightning.demos.boring_classes import DemoModel, BoringDataModule

class LitCLI(LightningCLI):
    def before_instantiate_classes(self) -> None:
        print(self.config[self.subcommand].trainer.logger)

def run():
    cli = LitCLI(
        DemoModel,
        BoringDataModule,
        trainer_defaults={"max_epochs": 2},
        save_config_callback=False,
    )


if __name__ == "__main__":
    run()
```

```console
python cli.py fit --trainer.logger WandbLogger
```

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -223,7 +223,7 @@ def __init__(self, *args, **kwarg):
 
     Args:
         name: Display name for the run.
-        save_dir: Path where data is saved (wandb dir by default).
+        save_dir: Path where data is saved.
```
Comment (tshu-w) [inline src/pytorch_lightning/loggers/wandb.py:226]: As discussed in https://github.com/Lightning-AI/lightning/issues/12036#issuecomment-1048268401, It's not a good idea to use [` wandb.Run.dir`](https://docs.wandb.ai/ref/python/run). In fact, currently WandbLogger uses the  `dir` pass to [`wandb.init`](https://docs.wandb.ai/ref/python/init) as `save_dir`, which is the same level of `wandb.Run.dir`, so I removed the confusing documentation.

File: src/pytorch_lightning/loggers/wandb.py [hunk 2]
```diff
@@ -304,7 +304,7 @@ def __init__(
             resume="allow",
             anonymous=("allow" if anonymous else None),
         )
-        self._wandb_init.update(**kwargs)
+        self._wandb_init.update(**{k: v for k, v in kwargs.items() if v is not None})
```
Comment (tshu-w) [inline src/pytorch_lightning/loggers/wandb.py:307]: Make sure `save_dir` is not overridden by `None` `dir`, see https://github.com/Lightning-AI/lightning/issues/14225

File: src/pytorch_lightning/loggers/wandb.py [hunk 3]
```diff
@@ -300,7 +301,7 @@ def __init__(
             name=name,
             project=project,
             id=version or id,
-            dir=save_dir,
+            dir=save_dir or kwargs.pop("dir"),
```
Comment (tshu-w) [inline src/pytorch_lightning/loggers/wandb.py:303]: Make sure `save_dir` is not overridden by `None` `dir`, see https://github.com/Lightning-AI/lightning/issues/14225.

File: tests/tests_pytorch/loggers/test_wandb.py [hunk 4]
```diff
@@ -58,7 +58,7 @@ def test_wandb_logger_init(wandb, monkeypatch):
     wandb.init.reset_mock()
     WandbLogger(project="test_project").experiment
     wandb.init.assert_called_once_with(
```
Comment (carmocca) [inline tests/tests_pytorch/loggers/test_wandb.py:60]: Can you add a test that does

```py
logger = WandbLogger(project="foo")
assert logger.save_dir is not None
```

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/loggers/test_wandb.py (hunk 4 above)
Comment (tshu-w) [reply tests/tests_pytorch/loggers/test_wandb.py:60]: added

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**